src/compiler/checker.ts

```diff
@@ -27681,6 +27681,8 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         if (!flowNode) {
             return declaredType;
         }
+        declaredType = isNoInferType(declaredType) ? (declaredType as SubstitutionType).baseType : declaredType;
+        initialType = isNoInferType(initialType) ? (initialType as SubstitutionType).baseType : initialType;

```

Comment:
My first instinct was that I should re-wrap the result of narrowing with `NoInfer` but quoting [the original `NoInfer` PR](https://github.com/microsoft/TypeScript/pull/56794):
> Other than blocking inference, NoInfer<T> markers have no effect on T. Indeed, T and NoInfer<T> are considered identical types in all other contexts.

So I don't see a reason why that would be necessary. I also tested something like this ([TS playground](https://www.typescriptlang.org/play?ts=5.5.0-dev.20240422&ssl=9&ssc=4&pln=1&pc=1#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXy1VRBgB4AVAPgApYBzALnnIEonyBuAKFElgRTpseeDmQYSFGlHYAaeGABGTWjEbwAcjgCSqRJKot4AXkrwAbjizA2Fq8G5cxEmNQDeSHDiYAiRbB94AF95VTojU3g3Lnh4ehj4AHpE+AA9AH4EsDwAZwx4OBzkCHzjAiISMJZuWOTY2IyuIOquIA)) to see if this would be needed and I don't think it is:
```ts
declare function inner<T>(arg: T): T;
declare function outer<T>(a: T, cb: (arg: NoInfer<T>) => void): void;

outer({ foo: "bar" }, (arg) => {
  arg
  // ^? (parameter) arg: NoInfer<{ foo: string; }>
  const result = inner(arg); // inferred despite `arg` being a `NoInfer` type!
  //    ^? const result: NoInfer<{ foo: string; }>
});
```

Reply:
I agree it's fine to erase `NoInfer` in variable, parameter, and property references, but I think it would be simpler and more consistent to do it first thing in the `getNarrowableTypeForReference` function. I think that's all you'd need.

Reply:
moved it to `getNarrowableTypeForReference`, works like a charm - thanks!